utils/sample_qc.py

```diff
@@ -0,0 +1,487 @@
+    pca_af_ht = unrelated_mt.annotate_rows(pca_af=hl.agg.mean(unrelated_mt.GT.n_alt_alleles()) / 2).rows()
+    pca_loadings = pca_loadings.annotate(pca_af=pca_af_ht[pca_loadings.key].pca_af)  # TODO: Evaluate if needed to write results at this point if relateds or not
+
+    if not related_samples_to_drop:
+        return pca_evals, pca_scores, pca_loadings
+    else:
+        pca_loadings = pca_loadings.persist()
+        pca_scores = pca_scores.persist()
+        related_mt = qc_mt.filter_cols(hl.is_defined(related_samples_to_drop[qc_mt.col_key]))
+        related_scores = pc_project(related_mt, pca_loadings)
+        pca_scores = pca_scores.union(related_scores)
+        return pca_evals, pca_scores, pca_loadings
+
+
+
+def compute_stratified_metrics_filter(ht: hl.Table, qc_metrics: List[str], strata: List[str] = None) -> hl.Table:
+    """
+    Compute median, MAD, and upper and lower thresholds for each metric used in pop- and platform-specific outlier filtering
+
+    :param MatrixTable ht: HT containing relevant sample QC metric annotations
+    :param list qc_metrics: list of metrics for which to compute the critical values for filtering outliers
+    :param list of str strata: List of annotations used for stratification. These metrics should be discrete types!
+    :return: Table grouped by pop and platform, with upper and lower threshold values computed for each sample QC metric
+    :rtype: Table
+    """
+
+    def make_pop_filters_expr(ht: hl.Table, qc_metrics: List[str]) -> hl.expr.SetExpression:
+        return hl.set(hl.filter(lambda x: hl.is_defined(x),
+                                [hl.or_missing(ht[f'fail_{metric}'], metric) for metric in qc_metrics]))
+
+    ht = ht.select(*strata, **ht.sample_qc.select(*qc_metrics)).key_by('s').persist()
+
+    def get_metric_expr(ht, metric):
+        return hl.bind(
+            lambda x: x.annotate(
+                upper=x.median + 4 * x.mad,
+                lower=x.median - 4 * x.mad
+            ),
+            hl.bind(
+                lambda elements, median: hl.struct(
+                    median=median,
+                    mad=1.4826 * hl.median(hl.abs(elements - median))
+                ),
+                *hl.bind(
+                    lambda x: hl.tuple([x, hl.median(x)]),
+                    hl.agg.collect(ht[metric])
+                )
+            )
+        )
+
+    agg_expr = hl.struct(**{metric: get_metric_expr(ht, metric) for metric in qc_metrics})
+    if strata:
+        ht = ht.annotate_globals(metrics_stats=ht.aggregate(hl.agg.group_by(hl.tuple([ht[x] for x in strata]), agg_expr)))
+    else:
+        ht = ht.annotate_globals(metrics_stats={(): ht.aggregate(agg_expr)})
+
+    strata_exp = hl.tuple([ht[x] for x in strata]) if strata else hl.tuple([])
+
+    fail_exprs = {
+        f'fail_{metric}':
+            (ht[metric] >= ht.metrics_stats[strata_exp][metric].upper) |
```

Comment:
How does this statement work? Doesn't `strata_exp = hl.tuple([ht[x] for x in strata])` give you `(ht.qc_platform, ht.qc_pop)` in the standard use case? But `metrics_stats` will be a dictionary where the keys are e.g., `('0', 'afr')` -- so how are you indexing into the `metrics_stats` with a generic `strata_exp`?

Reply:
I'm not totally sure what's confusing here. Your understanding of the code is entirely correct. So when running `ht = ht.transmute(**fail_exprs)`, `strata_exp` is evaluated for each row in `ht` and will evaluate to a tuple like `('0', 'afr')`, then the filter expressions (e.g. `(ht[metric] >= ht.metrics_stats[strata_exp][metric].upper)`) will pull the correct value in the dict using this `tuple` as key.
Apologies if I'm missing something..

Reply:
@gtiao : Just checking that I'm not missing anything here...

Reply:
OK, I think I got confused by the `strata_exp` statement and didn't realize that it gets evaluated. Thanks for walking me through that!